### PR TITLE
improve kubectl create --validate, aggregate errors and handle requir…

### PR DESCRIPTION
**Pull request by feihujiang**
improve kubectl create --validate, output aggregated errors and handle required fields

### CODE REVIEW

**Comment by feihujiang** [comment]
@satnam6502 @brendandburns @thockin Could you please review this PR?

File: pkg/api/validation/schema.go [hunk 1]
```diff
@@ -102,11 +109,18 @@ func (s *SwaggerSchema) ValidateObject(obj interface{}, apiVersion, fieldName, t
 	}
 	fields, ok := obj.(map[string]interface{})
 	if !ok {
-		return fmt.Errorf("expected object of type map[string]interface{} as value of %s field", fieldName)
+		return append(allErrs, fmt.Errorf("expected object of type map[string]interface{} as value of %s field", fieldName))
```
Comment (satnam6502) [inline pkg/api/validation/schema.go]: Why not report the actual type with `%T` ?

File: pkg/api/validation/schema.go [hunk 2]
```diff
@@ -102,11 +109,18 @@ func (s *SwaggerSchema) ValidateObject(obj interface{}, apiVersion, fieldName, t
...
 	}
 	if len(fieldName) > 0 {
 		fieldName = fieldName + "."
 	}
+	//handle required fields
+	for _, requiredKey := range model.Required {
+		_, ok := fields[requiredKey]
```
Comment (satnam6502) [inline pkg/api/validation/schema.go]: if _, ok := ... ; !ok {
...
}

Comment (hurf) [inline pkg/api/validation/schema.go]: I think having 'if' in a seperate line provides better readabiliy. Is there any other difference in these two formats?

Comment (satnam6502) [inline pkg/api/validation/schema.go]: I think having the if in this form makes for better readability.
It also constrains the scope of the `ok` variable to a narrower scope.

Comment (satnam6502) [inline pkg/api/validation/schema.go]: Also, this change also makes the code more consistent with code elsewhere in Kubernetes.

**Comment by feihujiang** [comment]
@satnam6502  OK, I changed according to your advice. Could you please review this PR again?

**Comment by satnam6502** [comment]
Please can you add a risk assessment for this PR? And link to a relevant bug? Thanks.

**Comment by satnam6502** [comment]
ok to test

**Comment by satnam6502** [comment]
@k8s-bot ok to test

**Comment by satnam6502** [comment]
Resting. If this fails again then please can someone investigate? Thanks.
@k8s-bot ok to test

**Comment by satnam6502** [comment]
@feihujiang please can you link this to an issue?

**Comment by satnam6502** [comment]
@k8s-bot ok to test

**Comment by satnam6502** [comment]
@feihujiang please can you look at why e2e fails?

**Comment by feihujiang** [comment]
#10079 @satnam6502  I post a new issues, and link this PR to it.

**Comment by feihujiang** [comment]
@satnam6502  When I opened Jenkins GCE e2e  Details,  it says "Access not possible". I think I have no right to open the Details now. Could you please post the Jenkins GCE e2e  Details for me? Thanks!

**Comment by satnam6502** [comment]
@k8s-bot ok to test

**Comment by satnam6502** [comment]
e2e passes now. Will wait for issue to be triaged to work out milestone.

**Comment by satnam6502** [comment]
Can you copy and paste some sample output? Thanks.

**Comment by feihujiang** [comment]
@satnam6502 OK 

The Invalid Pod:
apiVersion: v1beta3
kind: Pod
metadata:
  labels:
    name: cassandra
  name: cassandra
spec:
  containers:
  - args:
    - /run.sh
    resources:
      limits:
        cpu: "1"      
    name: cassandra
    ports:
    - name: cql
    - name: thrift
      containerPort: 9160
    volumeMounts:
    - name: data
      mountPath: /cassandra_data
    env:
    - name: MAX_HEAP_SIZE
      value: 512M
    - name: HEAP_NEWSIZE
      value: 100M
  volumes:
    - name: data
      emptyDir: {}

kubectl create --validate -f ./v1beta3/invalidCassandra.yaml
E0619 03:36:32.260674   14617 schema.go:153] Validation failed for: ports, [map[name:cql] map[containerPort:9160 name:thrift]]
E0619 03:36:32.260773   14617 schema.go:153] Validation failed for: containers, [map[volumeMounts:[map[mountPath:/cassandra_data name:data]] args:[/run.sh] env:[map[name:MAX_HEAP_SIZE value:512M] map[value:100M name:HEAP_NEWSIZE]] name:cassandra ports:[map[name:cql] map[containerPort:9160 name:thrift]] resources:map[limits:map[cpu:1]]]]
error validating "./v1beta3/invalidCassandra.yaml": error validating data: [ field image: is required, field containerPort: is required]

**Comment by satnam6502** [comment]
Thank you.

**Comment by satnam6502** [comment]
I will wait for the appropriate TL / meeting to determine the milestone for the associated issue before deciding what to do re: merging.

**Pull request merged**